tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,268 @@
+		strconv.Itoa(chainA.LatestCodeID),
+		`{"count": "0"}`, "",
+		initialization.ValidatorWalletName)
+
+	contracts, err := nodeA.QueryContractsFromID(chainA.LatestCodeID)
+	s.NoError(err)
+	s.Len(contracts, 1, "Wrong number of contracts for the counter")
+	contractAddr := contracts[0]
+
+	transferAmount := sdk.NewInt(10000000)
+	validatorAddr := nodeB.GetWallet(initialization.ValidatorWalletName)
+	nodeB.SendIBCTransfer(validatorAddr, contractAddr, fmt.Sprintf("%duluna", transferAmount.Int64()),
+		fmt.Sprintf(`{"wasm":{"contract":"%s","msg": {"increment": {}} }}`, contractAddr))
+
+	// check the balance of the contract
+	s.Eventually(func() bool {
+		balance, err := nodeA.QueryBalances(contractAddr)
+		s.Require().NoError(err)
+		if len(balance) == 0 {
+			return false
+		}
+		return balance[0].Amount.Equal(transferAmount)
+	},
+		1*time.Minute,
+		10*time.Millisecond)
+
+	// sender wasm addr
+	// senderBech32, err := ibchookskeeper.DeriveIntermediateSender("channel-0", validatorAddr, "terra")
+	var response interface{}
+	response, err = nodeA.QueryWasmSmart(contractAddr, `{"get_total_funds": {}}`)
+	s.Require().NoError(err)
+	fmt.Println("response: ", response)
+
+	s.Eventually(func() bool {
+		response, err = nodeA.QueryWasmSmart(contractAddr, `{"get_total_funds": {}}`)
+		if err != nil {
+			return false
+		}
+
+		fmt.Println("response: ", response)
+		totalFunds := response.([]interface{})[0]
+		amount, err := strconv.ParseInt(totalFunds.(map[string]interface{})["amount"].(string), 10, 64)
+		if err != nil {
+			return false
+		}
+		denom := totalFunds.(map[string]interface{})["denom"].(string)
+		// check if denom is luna token ibc
+		return sdk.NewInt(amount).Equal(transferAmount) && denom == initialization.TerraIBCDenom
+	},
+		10*time.Second,
+		10*time.Millisecond,
+	)
+}
+
+func (s *IntegrationTestSuite) TestPacketForwardMiddleware() {
+	if s.skipIBC {
+		s.T().Skip("Skipping Packet Forward Middleware tests")
+	}
+	chainA := s.configurer.GetChainConfig(0)
+	chainB := s.configurer.GetChainConfig(1)
+	chainC := s.configurer.GetChainConfig(2)
+
+	nodeA, err := chainA.GetDefaultNode()
+	s.NoError(err)
+	nodeB, err := chainB.GetDefaultNode()
+	s.NoError(err)
+	nodeC, err := chainC.GetDefaultNode()
+	s.NoError(err)
+
+	transferAmount := sdk.NewInt(10000000)
+
+	validatorAddr := nodeA.GetWallet(initialization.ValidatorWalletName)
+	s.Require().NotEqual(validatorAddr, "")
+
+	receiver := nodeB.GetWallet(initialization.ValidatorWalletName)
+
+	// query old bank balances
+	balanceReceiverOld, err := nodeC.QueryBalances(receiver)
+	s.NoError(err)
+	found, _ := balanceReceiverOld.Find("luna")
+	s.False(found)
+
+	nodeA.SendIBCTransfer(validatorAddr, receiver, fmt.Sprintf("%duluna", transferAmount.Int64()),
+		fmt.Sprintf(`{"forward":{"receiver":"%s","port":"transfer","channel":"channel-2"}}`, receiver))
+
+	// wait for ibc cycle
+	time.Sleep(30 * time.Second)
+
+	s.Eventually(func() bool {
+		balanceReceiver, err := nodeC.QueryBalances(receiver)
```

Comment:
please use at least terra (uluna)  -> terra -> cosmos and check ibc hash of uluna at final destination cosmos.

With fixed chain id, channel, port, and denom, the result of ibc hash of uluna at cosmos destination should be deterministic. This is a more necessary check than balance considering that wrong ibc hash will lead to different denom.

Reply:
<img width="1128" alt="Screenshot 2024-02-01 at 01 18 29" src="https://github.com/classic-terra/core/assets/23058164/c386105d-2a74-4b23-b187-ad291b6c818f">

currently in test, it is terra -> terra -> terra

Reply:
Ibc denom is derived based on port id, channel id and base denom not chain-id, not related to chain-id.

Reply:
thanks you, please continue with checking ibc hash